### PR TITLE
Better check for packages availability

### PR DESCRIPTION
**Pull request by apbard**
Following up huggingface/accelerate#1356
Refactor checks to avoid boilerplate and to ensure we are not picking up a folder that happens to be called as the package.

I assume all _*_available are for caching purposes. But is that really needed?

### CODE REVIEW

File: src/transformers/utils/import_utils.py [hunk 1]
```diff
@@ -35,6 +35,39 @@
...
+
+
+@overload
+def _is_package_available(pkg_name: str, return_version: Literal[False]) -> bool:
+    ...
+
+
+@overload
+def _is_package_available(pkg_name: str, return_version: Literal[True]) -> Tuple[bool, str]:
+    ...
```
Comment (sgugger) [inline src/transformers/utils/import_utils.py:51]: We don't do this in Transformers. The full signature of the function is way more than enough for us (also this is a private method, so no one will look at its signature).

**Review by sgugger — COMMENTED** [review]
Thanks a lot for the refactor, this is super nice!

There are a couple of packages that do not properly implement metadata (I kno for sure `opencv` does not since we are adding it in another PR), could you quickly check that all the packages for which you did this PR do implement the metadata? if they don't we need to rely on the old way, which is fine as it should be an exceptional case.

**Comment by apbard** [comment]
> Thanks a lot for the refactor, this is super nice!
> 
> There are a couple of packages that do not properly implement metadata (I kno for sure `opencv` does not since we are adding it in another PR), could you quickly check that all the packages for which you did this PR do implement the metadata? if they don't we need to rely on the old way, which is fine as it should be an exceptional case.

I have checked the packages and found issues only on sklearn (the import is sklearn and package scikit-learn) and decord.

Could you please double check: smdistributed, tensorflow_text, torchdistx?

**Comment by sgugger** [comment]
I have validated that `tensorflow_text` works. Installing `torchdistx` seems to painful for the time I have right now, but looking at the GitHub, everything should be fine.

Checking internally for `smdistributed` since it only exists in SageMaker envirnoments.

**Comment by sgugger** [comment]
Not hearing anything back on `smdistributed`, so let's just merge this PR and see if anyone complains :sweat_smile: 

Can you just fix the conflict?

File: src/transformers/utils/import_utils.py [hunk 2]
```diff
@@ -44,26 +62,83 @@
...
+
+_accelerate_available, _accelerate_version = _is_package_available("accelerate", return_version=True)
+_apex_available = _is_package_available("apex")
+_bitsandbytes_available = _is_package_available("bitsandbytes")
+_bs4_available = _is_package_available("bs4")
+_coloredlogs_available = _is_package_available("coloredlogs")
+_datasets_available = _is_package_available("datasets")
+_decord_available = importlib.util.find_spec("decord") is not None
+_detectron2_available = _is_package_available("detectron2")
+_diffusers_available = _is_package_available("diffusers")
```
Comment (sgugger) [inline src/transformers/utils/import_utils.py:77]: We don't have this check currently on main.

File: src/transformers/utils/import_utils.py [hunk 3]
```diff
@@ -44,26 +62,83 @@
...
+_ipex_available, _ipex_version = _is_package_available("intel_extension_for_pytorch", return_version=True)
+_jieba_available = _is_package_available("jieba")
+_kenlm_available = _is_package_available("kenlm")
+_keras_nlp_available = _is_package_available("keras_nlp")
+_librosa_available = _is_package_available("librosa")
+_natten_available = _is_package_available("natten")
+_ninja_available = _is_package_available("ninja")
+_onnx_available = _is_package_available("onnx")
+_openai_available = _is_package_available("openai")
+_opencv_available = importlib.util.find_spec("cv2") is not None
```
Comment (sgugger) [inline src/transformers/utils/import_utils.py:89]: Same here

File: src/transformers/utils/import_utils.py (hunk 2 above)
Comment (apbard) [reply src/transformers/utils/import_utils.py:77]: they have been introduced with #23214. should we remove them?
btw. I haven't checked that all "is_available" are used. just refactored assuming they are

Comment (sgugger) [reply src/transformers/utils/import_utils.py:77]: The private variables should have been cleaned up, this is a mistake indeed.

File: src/transformers/utils/import_utils.py [hunk 4]
```diff
@@ -333,10 +329,6 @@ def is_tf_available():
     return _tf_available
 
 
-def is_coloredlogs_available():
-    return _coloredlogs_available
```
[the remark this answers is not in the document]
Comment (apbard) [reply src/transformers/utils/import_utils.py]: reverted

**Comment by apbard** [comment]
don't think failures are due to this PR. Also main is failing

**Review by sgugger — APPROVED** [review]
Thanks!

**Pull request merged**